File: oofem-csharp/oofem-visualizer/Visualizer.cs
using BoDraw;
using static System.Math;

public class Visualizer
{
    public Structure Structure;

    public double ForceScale = 1;
    public double ConstraintSize = 1;
    public double ElementForceScale = 1;
    public double DisplacementScale = 1;

    public Visualizer(Structure structure)
    {
        this.Structure = structure;
    }

    public void DrawSystem(IBoDraw bd)
    {
        this.UpdateParameters();

        // Elements
        foreach (Element element in this.Structure.Elements)
        {
            Vector p1 = element.Node1.Position;
            Vector p2 = element.Node2.Position;

            bd.Add(new Line(p1[0], p1[1], p2[0], p2[1]).WithThickness(3));
        }

        // Constraints
        double a = this.ConstraintSize / 2;
        double b = 0.886 * this.ConstraintSize;

        foreach (Node node in this.Structure.Nodes)
        {
            double x1 = node.Position[0];
            double x2 = node.Position[1];

            if (node.Constraint.Fixed[0])
            {
                bd.Add(new Polygon(x1, x2, x1 - b, x2 - a, x1 - b, x2 + a));
            }
            if (node.Constraint.Fixed[1])
            {
                bd.Add(new Polygon(x1, x2, x1 - a, x2 - b, x1 + a, x2 - b));
            }
        }

        // Forces
        foreach (Node node in this.Structure.Nodes)
        {
            var f = node.Force.Components;

            if (f.Norm(2) != 0)
            {
                var p1 = node.Position;
                var p2 = p1 + this.ForceScale * f;

                bd.Add(new Arrow(p1[0], p1[1], p2[0], p2[1]).WithColor(Colors.Red).WithThickness(1.5));
            }
        }
    }

    public void DrawDeformation(BoDrawApp bd)
    {
        this.UpdateParameters();
        foreach (Element element in this.Structure.Elements)
        {
            Vector p1 = element.Node1.Position;
            Vector p2 = element.Node2.Position;
            Vector pu1 = p1 + this.DisplacementScale * element.Node1.Displacement;
            Vector pu2 = p2 + this.DisplacementScale * element.Node2.Displacement;

            bd.Add(new Line(p1[0], p1[1], p2[0], p2[1]).WithThickness(1));
            bd.Add(new Line(pu1[0], pu1[1], pu2[0], pu2[1]).WithThickness(3).WithColor(Colors.DarkBlue));
        }
    }

    public void DrawElementForces(BoDrawApp bd)
    {
        this.UpdateParameters();
        foreach (Element element in this.Structure.Elements)
        {
            double N = element.NormalForce();
            Vector d = element.D();
            Vector n = Vector.Build.DenseOfArray([d[1], -d[0]]);
            Vector p1 = element.Node1.Position;
            Vector p2 = element.Node2.Position;
            Vector p3 = p2 + this.ElementForceScale * N * n;
            Vector p4 = p1 + this.ElementForceScale * N * n;

            Polygon p = new Polygon(p1[0], p1[1], p2[0], p2[1], p3[0], p3[1], p4[0], p4[1]);
            if (N >= 0) { p.FillColor = Colors.Blue; }
            else { p.FillColor = Colors.Red; }
            p.FillOpacity = 0.3;
            bd.Add(p);
            bd.Add(new Line(p1[0], p1[1], p2[0], p2[1]).WithThickness(3));
        }
    }

    public void UpdateParameters()
    {
        double s = this.SystemSize();
        double maxF = this.Structure.Nodes.Max(n => n.Force.Components.Norm(2));
        double maxU = this.Structure.Nodes.Max(n => n.Displacement.Norm(2));
        double minN = this.Structure.Elements.Min(e => e.NormalForce());
        double maxN = this.Structure.Elements.Max(e => e.NormalForce());

        this.ConstraintSize = 0.03 * s;
        this.ForceScale = 0.2 / maxF * s;
        this.DisplacementScale = 0.1 * s / maxU;
        this.ElementForceScale = 0.3 * s / (maxN - minN);
    }

    public double SystemSize()
    {
        double[,] ranges = new double[,] {
             { double.MaxValue, double.MinValue },
             { double.MaxValue, double.MinValue }
        };

        foreach (Node node in this.Structure.Nodes)
        {
            for (int i = 0; i < 2; i++)
            {
                double p = node.Position[i];
                ranges[i, 0] = Min(ranges[i, 0], p);
                ranges[i, 1] = Max(ranges[i, 1], p);
            }
        }

        double d1 = ranges[0, 1] - ranges[0, 0];
        double d2 = ranges[1, 1] - ranges[1, 0];

        return Sqrt(d1 * d2);
    }
}

File: oofem-python/oofem-visualizer/visualizer.py
import numpy as np
import pyvista as pv


class Visualizer:

    def __init__(self, structure):

        self.structure = structure

        # Model visualization
        self.element_radius = 0.02
        self.node_radius = 0.04

        self.constraint_scale = 0.15
        self.force_scale = 0.25
        self.force_radius = 0.015

        # Result visualization
        self.undeformed_radius = 0.01
        self.deformed_radius = 0.025
        self.displacement_scale = 1.0

    def draw_elements(self, plotter):

        for element in self.structure.elements:

            point1 = element.node1.position
            point2 = element.node2.position

            line = pv.Line(point1, point2)

            plotter.add_mesh(
                line.tube(radius=self.element_radius),
                color="gray",
            )

    def draw_nodes(self, plotter):

        for node in self.structure.nodes:

            sphere = pv.Sphere(
                radius=self.node_radius,
                center=node.position,
            )

            plotter.add_mesh(
                sphere,
                color="black",
            )

    def draw_constraints(self, plotter):

        coordinate_directions = [
            np.array([1.0, 0.0, 0.0]),
            np.array([0.0, 1.0, 0.0]),
            np.array([0.0, 0.0, 1.0]),
        ]

        for node in self.structure.nodes:

            for i in range(3):

                if node.constraint.fixed[i]:

                    direction = coordinate_directions[i]

                    cone_center = (
                        node.position
                        - 0.5
                        * self.constraint_scale
                        * direction
                    )

                    cone = pv.Cone(
                        center=cone_center,
                        direction=direction,
                        height=self.constraint_scale,
                        radius=0.5 * self.constraint_scale,
                    )

                    plotter.add_mesh(
                        cone,
                        color="blue",
                    )

    def draw_forces(self, plotter):

        for node in self.structure.nodes:

            force = node.force.components
            force_magnitude = np.linalg.norm(force)

            if force_magnitude > 0.0:

                direction = force / force_magnitude

                arrow_length = (
                    self.force_scale
                    * force_magnitude
                )

                arrow_start = (
                    node.position
                    - arrow_length * direction
                )

                arrow = pv.Arrow(
                    start=arrow_start,
                    direction=direction,
                    scale=arrow_length,
                    shaft_radius=self.force_radius,
                    tip_radius=2.0 * self.force_radius,
                )

                plotter.add_mesh(
                    arrow,
                    color="red",
                )

    def draw_model(self):

        plotter = pv.Plotter()

        self.draw_elements(plotter)
        self.draw_nodes(plotter)
        self.draw_constraints(plotter)
        self.draw_forces(plotter)

        plotter.add_axes()
        plotter.enable_anti_aliasing()
        plotter.view_xy()

        plotter.show()

    def _draw_undeformed_structure(self, plotter):

        for element in self.structure.elements:

            point1 = element.node1.position
            point2 = element.node2.position

            line = pv.Line(
                point1,
                point2,
            )

            tube = line.tube(
                radius=self.undeformed_radius
            )

            plotter.add_mesh(
                tube,
                color="gray",
                opacity=0.45,
            )

    def _draw_deformed_structure(self, plotter):

        for element in self.structure.elements:

            deformed_point1 = (
                element.node1.position
                + self.displacement_scale
                * element.node1.displacement
            )

            deformed_point2 = (
                element.node2.position
                + self.displacement_scale
                * element.node2.displacement
            )

            line = pv.Line(
                deformed_point1,
                deformed_point2,
            )

            tube = line.tube(
                radius=self.deformed_radius
            )

            plotter.add_mesh(
                tube,
                color="blue",
            )

    def draw_deformed_structure(self):

        plotter = pv.Plotter()

        self._draw_undeformed_structure(plotter)
        self._draw_deformed_structure(plotter)

        self.draw_nodes(plotter)
        self.draw_constraints(plotter)
        self.draw_forces(plotter)

        plotter.add_axes()
        plotter.enable_anti_aliasing()
        plotter.view_xy()

        plotter.show()

    def _draw_element_forces(self, plotter):

        normal_forces = np.array(
            [
                element.normal_force()
                for element in self.structure.elements
            ],
            dtype=float,
        )

        if len(normal_forces) == 0:
            return None

        maximum_force = np.max(
            np.abs(normal_forces)
        )

        if maximum_force < 1e-12:
            maximum_force = 1.0

        color_limits = [
            -maximum_force,
            maximum_force,
        ]

        normal_force_mapper = None

        for element in self.structure.elements:

            deformed_point1 = (
                element.node1.position
                + self.displacement_scale
                * element.node1.displacement
            )

            deformed_point2 = (
                element.node2.position
                + self.displacement_scale
                * element.node2.displacement
            )

            line = pv.Line(
                deformed_point1,
                deformed_point2,
            )

            normal_force = float(
                element.normal_force()
            )

            line.point_data["Normal force"] = np.array(
                [
                    normal_force,
                    normal_force,
                ],
                dtype=float,
            )

            tube = line.tube(
                radius=self.deformed_radius
            )

            actor = plotter.add_mesh(
                tube,
                scalars="Normal force",
                cmap="bwr",
                clim=color_limits,
                show_scalar_bar=False,
            )

            if normal_force_mapper is None:
                normal_force_mapper = actor.mapper

        return normal_force_mapper

    def draw_element_forces(self):

        plotter = pv.Plotter()

        self._draw_undeformed_structure(plotter)

        normal_force_mapper = (
            self._draw_element_forces(plotter)
        )

        self.draw_nodes(plotter)
        self.draw_constraints(plotter)
        self.draw_forces(plotter)

        if normal_force_mapper is not None:

            plotter.add_scalar_bar(
                title="Normal force [N]",
                mapper=normal_force_mapper,
                n_labels=5,
                fmt="%.2e",
                vertical=False,
            )

        plotter.add_axes()
        plotter.enable_anti_aliasing()
        plotter.view_xy()

        plotter.show()